wasm/src/object.c

```diff
@@ -71,6 +74,251 @@ static void __copy_object_elem(opa_object_t *result, opa_value *a, opa_value *b)
+                }
+                case OPA_OBJECT:
+                {
+                   node = opa_cast_object(child);
+                }
+            }
+        }
+
+        if (opa_value_compare(done, opa_boolean(false)) == 0)
+        {
+            opa_value *key = terms->elems[terms->len - 1].v;
+            opa_object_insert(node, key, opa_null());
+        }
+    }
+
+    return root;
+
+}
+
+opa_array_t *__parse_path(opa_value *a)
+{
+    // paths can either be a `/` separated json path or
+    // an array or set of values
+
+    opa_array_t *path_segments = opa_cast_array(opa_array());
+
+    switch (opa_value_type(a))
+    {
+        case OPA_STRING:
+        {
+            opa_string_t *x = opa_cast_string(a);
+
+            if (x->len == 0)
+            {
+                return path_segments;
+            }
+
+            opa_value *s = opa_strings_split(opa_strings_trim_left(a, opa_string_terminated("/")), opa_string_terminated("/"));
+
+            opa_array_t *parts = opa_cast_array(s);
+
+            for (int i = 0; i < parts->len; i++)
+            {
+                opa_value *s = opa_strings_replace(parts->elems[i].v, opa_string_terminated("~1"), opa_string_terminated("/"));
+                opa_value *part = opa_strings_replace(s, opa_string_terminated("~0"), opa_string_terminated("~"));
+                opa_array_append(path_segments, part);
+            }
+
+            return path_segments;
+        }
+        case OPA_ARRAY:
+        {
+            opa_array_t *y = opa_cast_array(a);
+
+            for (int i = 0; i < y->len; i++)
+            {
+                opa_array_append(path_segments, y->elems[i].v);
+            }
+
+            return path_segments;
+        }
+    }
+
+    return NULL;
+}
+
+opa_array_t *__get_json_paths(opa_value *a)
+{
+    opa_array_t *paths = opa_cast_array(opa_array());
+
+    switch (opa_value_type(a))
+    {
+        case OPA_SET:
+        {
+            opa_set_t *x = opa_cast_set(a);
+
+            for (int i = 0; i < x->n; i++)
+            {
+                opa_set_elem_t *elem = x->buckets[i];
+
+                while (elem != NULL)
+                {
+                    opa_array_t* path = __parse_path(elem->v);
+
+                    if (path == NULL)
+                    {
+                       return NULL;
+                    }
+                    opa_array_append(paths, &path->hdr);
+                    elem = elem->next;
+                }
+            }
+
+            return paths;
+        }
+        case OPA_ARRAY:
+        {
+            opa_array_t *y = opa_cast_array(a);
+
+            for (int i = 0; i < y->len; i++)
+            {
+                opa_array_t* path = __parse_path(y->elems[i].v);
+
+                if (path == NULL)
+                {
+                    return NULL;
+                }
+                opa_array_append(paths, &path->hdr);
+            }
+
+            return paths;
+        }
+    }
```

Comment:
We should be able to use `opa_value_iter` here instead of having specific cases for sets and objects. You can find examples of how `opa_value_iter` is called in value.c.

Reply:
Done